subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java

```diff
@@ -117,68 +53,35 @@ public Object unpackState() {
-        // TODO - should hold the project lock to do the isolation. Should work the same way as artifact transforms (a work node does the isolation, etc)
-        P isolatedParameters = isolatableFactory.isolate(parameters).isolate();
-        // TODO - reuse this in other places
-        ServiceLookup instantiationServices = instantiationServicesFor(isolatedParameters);
-        try {
-            return Try.successful(instantiate(instantiationServices));
-        } catch (Exception e) {
-            return Try.failure(instantiationException(e));
-        }
+    public void maybeStop() {
+        // TODO
     }
 
-    private ServiceLifecycleException instantiationException(Exception e) {
-        return new ServiceLifecycleException("Failed to create service '" + name + "'.", e);
+    @SuppressWarnings("unused") // Used via instrumentation
+    public static <P extends BuildServiceParameters, T extends BuildService<P>> void setBuildServiceAsConvention(DefaultProperty<T> property, ServiceRegistry serviceRegistry, String buildServiceName) {
+        ConsumedBuildServiceProvider<T, P> lazyProvider = new ConsumedBuildServiceProvider<>(buildServiceName, property.getType(), serviceRegistry);
+        property.convention(lazyProvider);
     }
 
-    private T instantiate(ServiceLookup instantiationServices) {
-        return instantiationScheme.withServices(instantiationServices).instantiator().newInstance(implementationType);
-    }
+    public abstract BuildServiceDetails<T, P> getServiceDetails();
 
-    private ServiceLookup instantiationServicesFor(@Nullable P isolatedParameters) {
-        return isolationScheme.servicesForImplementation(
-            isolatedParameters,
-            internalServices,
-            ImmutableList.of(),
-            serviceType -> false
-        );
-    }
+    public abstract String getName();
 
-    @Override
-    public ExecutionTimeValue<? extends T> calculateExecutionTimeValue() {
-        return ExecutionTimeValue.changingValue(this);
-    }
-
-    public void maybeStop() {
-        synchronized (this) {
-            if (instance != null) {
-                instance.ifSuccessful(t -> {
-                    if (t instanceof AutoCloseable) {
-                        try {
-                            ((AutoCloseable) t).close();
-                        } catch (Exception e) {
-                            throw new ServiceLifecycleException("Failed to stop service '" + name + "'.", e);
-                        }
-                    }
-                });
-            }
+    /**
+     * Are the given providers referring to the same service provider?
+     *
+     * This method does not distinguish between consumed/registered providers.
+     */
+    public static boolean isSameService(Provider<? extends BuildService<?>> thisProvider, Provider<? extends BuildService<?>> anotherProvider) {
+        if (!(thisProvider instanceof BuildServiceProvider && anotherProvider instanceof BuildServiceProvider)) {
```

Comment:
❓ Would it make sense to have a cheap reference equality check first?

Reply:
Added a second if after the first if. We still want to consider only BSPs (instead of regular Providers).